mlflow/gateway/providers/palm.py

```diff
@@ -0,0 +1,221 @@
+                    {
+                        "message": {
+                            "role": c["author"],
+                            "content": c["content"],
+                        },
+                        "metadata": {},
+                    }
+                    for c in resp["candidates"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        key_mapping = {
+            "stop": "stopSequences",
+            "candidate_count": "candidateCount",
+            "max_tokens": "maxOutputTokens",
+        }
+        for k1, k2 in key_mapping.items():
+            if k2 in payload:
+                raise HTTPException(
+                    status_code=422, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+        payload = rename_payload_keys(payload, key_mapping)
+        payload["prompt"] = {"text": payload["prompt"]}
+        resp = await self._request(
+            f"{self.config.model.name}:generateText",
+            payload,
+        )
+        # Response example (https://developers.generativeai.google/api/rest/generativelanguage/models/generateText)
+        # ```
+        # {
+        #   "candidates": [
+        #     {
+        #       "output": "Once upon a time, there was a young girl named Lily...",
+        #       "safetyRatings": [
+        #         {
+        #           "category": "HARM_CATEGORY_DEROGATORY",
+        #           "probability": "NEGLIGIBLE"
+        #         }, ...
+        #       ]
+        #     {
+        #       "output": "Once upon a time, there was a young boy named Billy...",
+        #       "safetyRatings": [
+        #           ...
+        #       ]
+        #     }
+        #   ]
+        # }
+        # ```
+        return completions.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "text": c["output"],
+                        "metadata": {},
+                    }
+                    for c in resp["candidates"]
+                ],
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def embeddings(self, payload: embeddings.RequestPayload) -> embeddings.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        postfix = "embedText"
+
+        # Remap text to texts to setup for batch embeddings
+        if isinstance(payload["text"], list):
+            postfix = "batchEmbedText"
```

Comment:
can we always use `batchEmbedText` to simplify the code?

Reply:
(unless there is a big performance gap)

Reply:
Yeah doesn't seem to be much of a performance impact so I updated to always use `batchEmbedText`.

Here's the single-input `embedText` performance:
<img width="1502" alt="Screenshot 2023-10-01 at 5 47 18 PM" src="https://github.com/databricks/universe/assets/87999496/450b2662-2790-4fa9-aab9-c9469db9e188">

And here's the `batchEmbedText` performance with a single-input:
<img width="1503" alt="Screenshot 2023-10-06 at 12 57 06 AM" src="https://github.com/mlflow/mlflow/assets/87999496/0ba21a3b-735e-4d8f-b5bf-0bca2c37da96">